### PR TITLE
Update deploy process

### PR DESCRIPTION
**Pull request by carolynjoneslee**
### What’s this PR do?
This PR makes a few updates to the deploy process, with the goal of allowing more thorough testing before each release.

Specifically, it allows us to deploy separately to 3 locations:
 1. the [CDN root](https://cdn.yesgraph.com/yesgraph-invites.js) (for production)
 2. the [ "dev/" branch of the CDN](https://cdn.yesgraph.com/dev/yesgraph-invites.js) (for testing changes live, before updating the root CDN), and
 3. to a local "deployed/" folder (for testing the deploy process itself, without actually updating the CDN).

### Where should the reviewer start?
Start by reviewing the updated deploy task in gulp/tasks/deploy.js

### How should this be manually tested?
Try running a test deploy! Use the `--test` flag to indicate that instead of pushing to S3 we should just save the updated files locally, and the `--prod` flag to indicate that we should process all files, (not just the dev/ files).

All in all the command is this:
`gulp deploy --test --prod`

You should now have a whole bunch of files in the new deployed/ folder:
```
deployed/
├── dev
│   ├── v0.0.6
│   │   ├── yesgraph-invites.css
│   │   └── yesgraph-invites.min.css
│   ├── v0.1.7
│   │   ├── yesgraph.js
│   │   ├── yesgraph.min.js
│   │   └── yesgraph.min.js.map
│   ├── v1.1.6
│   │   ├── yesgraph-invites.js
│   │   ├── yesgraph-invites.min.js
│   │   └── yesgraph-invites.min.js.map
│   ├── yesgraph-invites.css
│   ├── yesgraph-invites.js
│   ├── yesgraph-invites.min.css
│   ├── yesgraph-invites.min.js
│   ├── yesgraph-invites.min.js.map
│   ├── yesgraph.js
│   ├── yesgraph.min.js
│   └── yesgraph.min.js.map
├── v0.0.6
│   ├── yesgraph-invites.css
│   └── yesgraph-invites.min.css
├── v0.1.7
│   ├── yesgraph.js
│   ├── yesgraph.min.js
│   └── yesgraph.min.js.map
├── v1.1.6
│   ├── yesgraph-invites.js
│   ├── yesgraph-invites.min.js
│   └── yesgraph-invites.min.js.map
├── yesgraph-invites.css
├── yesgraph-invites.js
├── yesgraph-invites.min.css
├── yesgraph-invites.min.js
├── yesgraph-invites.min.js.map
├── yesgraph.js
├── yesgraph.min.js
└── yesgraph.min.js.map

7 directories, 32 files
```
### What are the relevant tickets?
[Update Superwidget deploy process](https://app.asana.com/0/59202558034519/162364139954367)

### Does the knowledge base need an update?
Yeah, I should add some documentation to the wiki.

### CODE REVIEW

File: gulp/config.js [hunk 1]
```diff
@@ -7,8 +7,8 @@ var dest = {
     dev: "./dist/dev",
     root: "./dist"
 };
-var __SDK_VERSION__ = "v0.1.6";
-var __SUPERWIDGET_VERSION__ = "v1.1.5";
+var __SDK_VERSION__ = "v0.1.7";
+var __SUPERWIDGET_VERSION__ = "v1.1.6";
```
Comment (carolynjoneslee) [inline gulp/config.js]: These new version numbers are consistent with the CDN, no actual change here

File: gulptools/tasks/deploy.js [hunk 2]
```diff
@@ -1,39 +0,0 @@
-var gulp = require("gulp");
```
Comment (carolynjoneslee) [inline gulptools/tasks/deploy.js]: This file got changed and renamed (to gulp/tasks/deploy.js), not deleted

File: gulptools/tasks/version.js [hunk 3]
```diff
@@ -1,83 +0,0 @@
-var gulp = require("gulp");
```
Comment (carolynjoneslee) [inline gulptools/tasks/version.js]: This file got changed and renamed (to gulp/tasks/version.js), not deleted

File: src/yesgraph-invites.js [hunk 4]
```diff
@@ -895,8 +895,8 @@
                 function getWidgetOptions() {
                     var d = $.Deferred();
                     var OPTIONS_ENDPOINT = '/apps/' + YesGraphAPI.app + '/js/get-options';
-                    // Retry failed request up to 3 times, waiting 300ms between tries
-                    YesGraphAPI.hitAPI(OPTIONS_ENDPOINT, "GET", {}, null, 3, 300).done(function (data) {
+                    // Retry failed request up to 3 times, waiting 1500ms between tries
+                    YesGraphAPI.hitAPI(OPTIONS_ENDPOINT, "GET", {}, null, 3, 1500).done(function (data) {
```
Comment (kendallc) [inline src/yesgraph-invites.js]: Why 300 or 1500 seconds? I wonder what the best number is here.

File: gulp/tasks/deploy.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+ */
+
+gulp.task("deploy", ["build"], function() {    
+    var config = require("../config");  // Load config after versioning
+    var devFiles = "dist/dev/**/*.@(js|css|map)";
+    var publisher = aws.create(config.s3);
+    var cloneSink = clone.sink();
+
+    validateDeployType();
+    var filesChanged = argv.dev ? "dev" : "all";
```
Comment (kendallc) [inline gulp/tasks/deploy.js]: Looks like we're not checking the "--prod" and "--test" command line inputs here, do we need them? Maybe we can simplify the input format somehow since it seems to be two boolean variables.

File: gulp/tasks/deploy.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+    var deployPrep = lazypipe()
+        .pipe(function(){
+            // We use cloneSink here to create a copy of the files and rename
+            // only the copies (so that they can be sorted by version in S3).
+            return cloneSink;
+        })
+        .pipe(rename, function(path){
+            // Add versions to the filepath
+            if (startsWith(path.basename, "yesgraph-invites") && path.extname == ".css") {
```
Comment (kendallc) [inline gulp/tasks/deploy.js]: Just a stylistic/programming comment: It might make sense to have few lists of input files (e.g. "css_file_list", "js_file_list") rather than have many of these regexes. I notice that we have a lot of these regexes being used to match the same handful of files.

Alternatively, maybe we could do regexes at the top of the file to get the list of css, js files and use those lists later on in the script. It might make the code more readable, regexes take some time for my brain to parse.

File: src/yesgraph-invites.js (hunk 4 above)
Comment (carolynjoneslee) [inline src/yesgraph-invites.js]: Yeah I'm not sure why this change is getting included here, we already have it on the CDN, so I probably just forgot to commit the change at the time.

We started out with a 300ms timeout, but it was still throwing errors, so I made it longer, to give the API a bit more time to fix itself. The errors stopped, so I think 1500ms is the right number!

File: gulp/tasks/deploy.js (hunk 5 above)
Comment (carolynjoneslee) [inline gulp/tasks/deploy.js]: Yeah one thing that might make it simpler is just assuming by default that it's a live deploy on all the files, and then only set it explicitly if we don't want the defaults:

Normal deploy, handling all files and updating the CDN:
`gulp deploy`

Test deploy, handling all files but not updating the CDN:
`gulp deploy --test`

Dev deploy, handling only dev files and not updating the CDN:
`gulp deploy --dev --test`

File: src/yesgraph-invites.js (hunk 4 above)
Comment (kendallc) [inline src/yesgraph-invites.js]: Okay, cool.

File: gulp/tasks/deploy.js (hunk 5 above)
Comment (kendallc) [inline gulp/tasks/deploy.js]: Yeah, having those defaults will make it a lot simpler, reducing the number of inputs

File: gulp/tasks/deploy.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+var superwidgetFiles = ["yesgraph-invites.js", "yesgraph-invites.min.js", "yesgraph-invites.min.js.map"];
+var cssFiles = ["yesgraph-invites.css", "yesgraph-invites.min.css"];
+
+/*
+ * This task handles the entire deploy process from start to finish,
+ * but executes different task sequences based on whether we're updating
+ * dev versions, prod versions, or simply running a test deploy locally.
+ *
+ * Options:
+ * --prod   update all files
```
Comment (kendallc) [inline gulp/tasks/deploy.js]: Might remove "--prod" and "--live" options here
